kerastuner/engine/base_tuner.py

```diff
@@ -0,0 +1,232 @@
+            and returns a Model instance).
+        directory: String. Path to the working directory (relative).
+        project_name: Name to use as prefix for files saved
+            by this Tuner.
+        logger: Optional. Instance of Logger class, used for streaming data
+            to Cloud Service for monitoring.
+        tuner_id: Optional. Used only with multi-worker DistributionStrategies.
+    """
+
+    def __init__(self,
+                 oracle,
+                 hypermodel,
+                 directory=None,
+                 project_name=None,
+                 logger=None,
+                 tuner_id=None):
+        if not isinstance(oracle, oracle_module.Oracle):
+            raise ValueError('Expected oracle to be '
+                             'an instance of Oracle, got: %s' % (oracle,))
+        self.oracle = oracle
+
+        if isinstance(hypermodel, hm_module.HyperModel):
+            self.hypermodel = hypermodel
+        else:
+            if not callable(hypermodel):
+                raise ValueError(
+                    'The `hypermodel` argument should be either '
+                    'a callable with signature `build(hp)` returning a model, '
+                    'or an instance of `HyperModel`.')
+            self.hypermodel = hm_module.DefaultHyperModel(hypermodel)
+
+        # To support tuning distribution.
+        self.tuner_id = tuner_id if tuner_id is not None else 0
+
+        # Ops and metadata
+        self.directory = directory or '.'
+        self.project_name = project_name or 'untitled_project'
+
+        # Logs etc
+        self.logger = logger
+        self._display = tuner_utils.Display()
+
+    def search(self, *fit_args, **fit_kwargs):
+        self.on_search_begin()
+        while True:
+            trial = self.oracle.create_trial(self.tuner_id)
+            if trial.status == trial_module.TrialStatus.STOPPED:
+                # Oracle triggered exit.
+                break
+            if trial.status == trial_module.TrialStatus.IDLE:
+                # Oracle is calculating, resend request.
+                continue
+
+            self.on_trial_begin(trial)
+            self.run_trial(trial, *fit_args, **fit_kwargs)
+            self.on_trial_end(trial)
+        self.on_search_end()
+
+    def run_trial(self, trial, *fit_args, **fit_kwargs):
+        raise NotImplementedError
+
+    def on_search_begin(self):
+        if self.logger:
+            self.logger.register_tuner(self.get_state())
+
+    def on_trial_begin(self, trial):
+        if self.logger:
+            self.logger.register_trial(trial.trial_id, trial.get_state())
+
+    def on_trial_end(self, trial):
+        self.oracle.end_trial(
+            trial.trial_id, trial_module.TrialStatus.COMPLETED)
+        self._display.on_trial_end(trial)
+        self._save_trial(trial)
+        self.save()
+
+    def on_search_end(self):
+        if self.logger:
+            self.logger.exit()
+
+    def get_best_models(self, num_models=1):
```

Comment:
Perhaps we should provide `save_model` and `load_model` to be overridden by the user, and a default implementation for `get_best_model` (as well as default model saving behavior in `on_trial_end`).

This would make to much easier to implement new basic tuners for sklearn, xgboost, or anything else (just override `run_trial`, `save_model` and `load_model`, which are just 2 lines each)

Reply:
SGTM, yeah I think that makes writing Tuners a lot nicer. Added this to `BaseTuner` and implemented in `Tuner`